server/sonar-server/src/main/java/org/sonar/server/platform/StartupMetadataProvider.java

```diff
@@ -0,0 +1,90 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.platform;
+
+import org.picocontainer.injectors.ProviderAdapter;
+import org.sonar.api.CoreProperties;
+import org.sonar.api.SonarQubeSide;
+import org.sonar.api.SonarRuntime;
+import org.sonar.api.ce.ComputeEngineSide;
+import org.sonar.api.server.ServerSide;
+import org.sonar.api.utils.System2;
+import org.sonar.core.util.UuidFactory;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.property.PropertiesDao;
+import org.sonar.db.property.PropertyDto;
+import org.sonar.server.platform.cluster.Cluster;
+
+import static com.google.common.base.Preconditions.checkState;
+import static org.apache.commons.lang.StringUtils.isBlank;
+
+@ComputeEngineSide
+@ServerSide
+public class StartupMetadataProvider extends ProviderAdapter {
+
+  private static final String PROPERTY_STARTED_AT = "sonar.core.startedAt";
+
+  private StartupMetadata cache = null;
+
+  public StartupMetadata provide(UuidFactory uuidFactory, System2 system, SonarRuntime runtime, Cluster cluster, DbClient dbClient) {
+    if (cache == null) {
+      if (runtime.getSonarQubeSide() == SonarQubeSide.SERVER && cluster.isStartupLeader()) {
+        cache = create(uuidFactory, system, dbClient);
+      } else {
+        cache = load(dbClient);
+      }
+    }
+    return cache;
+  }
+
+  /**
+   * Generate and persist in database
+   */
+  private StartupMetadata create(UuidFactory uuidFactory, System2 system, DbClient dbClient) {
+    String startupId = uuidFactory.create();
+    long startedAt = system.now();
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      PropertiesDao dao = dbClient.propertiesDao();
+      dao.insertProperty(dbSession, new PropertyDto().setKey(CoreProperties.SERVER_ID).setValue(startupId));
+      dao.insertProperty(dbSession, new PropertyDto().setKey(PROPERTY_STARTED_AT).setValue(String.valueOf(startedAt)));
+      dbSession.commit();
+    }
+    return new StartupMetadata(startupId, startedAt);
+  }
+
+  /**
+   * Load from database
+   */
+  private StartupMetadata load(DbClient dbClient) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      String startupId = selectProperty(dbClient, dbSession, CoreProperties.SERVER_ID);
+      String startedAt = selectProperty(dbClient, dbSession, PROPERTY_STARTED_AT);
+      return new StartupMetadata(startupId, Long.parseLong(startedAt));
+    }
+  }
+
+  private String selectProperty(DbClient dbClient, DbSession dbSession, String key) {
+    PropertyDto prop = dbClient.propertiesDao().selectGlobalProperty(dbSession, key);
+    checkState(prop != null, "Property %s is missing in database", key);
+    checkState(!isBlank(prop.getValue()), "Property %s is set but empty in database", key);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) NullPointerException might be thrown as 'prop' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)